src/main/java/org/openpnp/machine/reference/ReferencePnpJobProcessor.java

```diff
@@ -291,7 +293,29 @@ private void prepMachine() throws JobProcessorException {
             // Discard any currently picked parts
             discardAll(head);
         }
-        
+
+        private void prepFeeders() throws JobProcessorException {
+            // Everything still looks good, so prepare the feeders.
+            fireTextStatus("Preparing feeders.");
+            Machine machine = Configuration.get().getMachine();
+            // Get all the feeders that are used in the pending placements.
+            for (Feeder feeder : machine.getFeeders()) {
+                if (feeder.isEnabled() && feeder.getPart() != null) {
+                    for (JobPlacement placement : getPendingJobPlacements()) {
+                        if (placement.getPartId() == feeder.getPart().getId()) {
+                            try {
```

Comment:
Hi @vonnieda 

Thanks for looking into this.

> Feeders can become enabled during a job run

I didn't know that. How does this happen? By user intervention?

Normally, the Job preparation is just a speed optimization i.e. things should work even when the preparation is not water tight. That's because the `feed()` operation will double-check here:

https://github.com/openpnp/openpnp/blob/cba67073f2c660f24bff4e91c48eb3e8e1bec35f/src/main/java/org/openpnp/machine/reference/feeder/BlindsFeeder.java#L329-L334

However, if none of the currently loaded nozzle tips are allowed for pushing, this will block the operation, because the nozzle tip will not be changed "on the fly" just to open the cover (the job processor already loaded the right nozzle tips, and we don't want to fumble with those). 

Changing the nozzle tip for one that is allowed for pushing is only done in the job preparation.  

See [org.openpnp.machine.reference.feeder.BlindsFeeder.nozzleForPushing(boolean)](https://github.com/openpnp/openpnp/blob/cba67073f2c660f24bff4e91c48eb3e8e1bec35f/src/main/java/org/openpnp/machine/reference/feeder/BlindsFeeder.java#L1162) and specifically the `loadNozzleTipIfNeeded` parameter, that is only set to `true` (indirectly through [org.openpnp.machine.reference.feeder.BlindsFeeder.actuateCover(boolean, boolean)](https://github.com/openpnp/openpnp/blob/cba67073f2c660f24bff4e91c48eb3e8e1bec35f/src/main/java/org/openpnp/machine/reference/feeder/BlindsFeeder.java#L1207)) in `org.openpnp.machine.reference.feeder.BlindsFeeder.prepareForJob()`.

https://github.com/openpnp/openpnp/blob/cba67073f2c660f24bff4e91c48eb3e8e1bec35f/src/main/java/org/openpnp/machine/reference/feeder/BlindsFeeder.java#L1195-L1201

Having to load a special "pushing" nozzle tip is the actual reason for doing the "open" all at once up front when the job starts. Using this scheme, users could even 3D-print a special "pushing" nozzle tip if they shy away from using their dear real ones. 

**So you're right, there is a problem.**

If the late enabling of feeders is due to user intervention, I wonder if there is a "resume paused job" event, that I could hook another feeder preparation into? 
(though I thought, restarting a job will go through the normal startup code?). 

If it is _not_ due to user intervention, then please explain when and why this happens.

Thanks a lot, 
_Mark

Reply:
Yes, it's due to user intervention. Simplest case is the user loads a feeder that was previously unloaded and clicks enable instead of just reloading another when it runs out.

In general, I think that this prepareForJob change is fragile. A feeder should always be ready to have feed() operations performed on it, as these also happen during setup and manually by the user at any time.

Perhaps you could just have some static members on your class to keep track of whether the feeders have been all opened or all closed and perform the action during the first feed of any of them? This is similar to how I handle the banks on the auto slot feeder.

Jason

Reply:
Ok, I see. 

Two problems:

1. I don't want to open all the feeders just because the user presses a manual feed on _one_ of them. I specifically **only** want the all-at-once-preparation **inside a job**. And I would like to restrict it to the parts that are actually used inside that job. So if I _did_ it inside `feed()` I'd need access to the `getPendingJobPlacements()` and the `JobProcessor` running state etc.. This does not sound right. A feeder should not have to worry about the `JobProcessor` at all. 

2. When the first `feed()` happens inside a job, the right nozzle tips for a series of placements have already been loaded. It would be inefficient, to then again exchange it for the "pushing" nozzle tip and back (even if this happens only once inside the Job). 

Another approach:

1. Keep the preparation as it is now. It is not watertight, but the "right place" to do it and generally beneficial for efficiency.  
2. Assumption 1: If the user enables a feeder in the middle of the Job (or adds/changes a placement) this is considered a rare and exceptional event where a speed penalty is acceptable.
3. Assumption 2: If the user performs a manual feed outside a job, this is a one-off operation that is also not critical for speed. 
4. Taking 2. and 3. into consideration, I'll change `feed()` so it will automatically change the nozzle tip as needed for pushing and then immediately back to the original nozzle tip to not upset the `JobProcessors`'s or user's working assumptions.
5. This way only the exceptional/manual case will suffer a speed penalty, while the routine case will run as efficiently as possible.

I'll try and do that now, as a proposal.

_Mark

Reply:
Ok, I see. 

Two problems:

1. I don't want to open all the feeders just because the user presses a manual feed on _one_ of them. I specifically **only** want the all-at-once-preparation **inside a job**. And I would like to restrict it to the parts that are actually used inside that job. So if I _did_ it inside `feed()` I'd need access to the `getPendingJobPlacements()` and the `JobProcessor` running state etc.. This does not sound right. A feeder should not have to worry about the `JobProcessor` at all. 

2. When the first `feed()` happens inside a job, the right nozzle tips for a series of placements have already been loaded. It would be inefficient, to then again exchange it for the "pushing" nozzle tip and back (even if this happens only once inside the Job). 

Another approach:

1. Keep the preparation more or less as it is now. It is not watertight, but the "right place" to do it and generally beneficial for efficiency.  
2. Assumption 1: If the user enables a feeder in the middle of the Job (or adds/changes a placement) this is considered a rare and exceptional event where a speed penalty is acceptable.
3. Assumption 2: If the user performs a manual feed outside a job, this is a one-off operation that is also not critical for speed. 
4. Taking 2. and 3. into consideration, I'll change `feed()` so it will automatically change the nozzle tip as needed for pushing and then immediately back to the original nozzle tip to not upset the `JobProcessors`'s or user's working assumptions.
5. This way, only the exceptional/manual case will suffer a speed penalty, while the routine case will run as efficiently as possible.

I'll try and do that now, as a proposal.

_Mark